Fix include/exclude in QgsRange

Djedouas opened the pull request:
## Description

### fix QgsRange::overlaps()

When comparing 2 QgsRange objects for overlap, only the first include/exclude limits restrictions were considered.

This leads to errors like this, where range1.overlaps(range2) != range2.overlaps(range1):

```python
<QgsIntRange: [4, 8)>   # left end not included
<QgsIntRange: [8, 10)>  # left end not included
r1 overlaps r2: False
r2 overlaps r1: True

<QgsIntRange: (4, 8)>  # both ends included
<QgsIntRange: [8, 10]>  # left end not included
r1 overlaps r2: False
r2 overlaps r1: True
```

### fix time limits include/exclude for vector layer properties

In **fixed mode** the include/exclude limits parameter was not used (always considered include both).

---

Co-funded by [CNES](https://cnes.fr) and QGIS.org

Djedouas committed:
fix(vector-temporal-range-properties): consider include/exclude end in fixed mode

For vector layers.

When temporal range is in fixed mode, include/exclude end is not
considered (always included).

In other temporal range modes, a QGIS expression is constructed which
already takes it into account.

Djedouas committed:
fix(qgsrange): consider include/exclude limits on both range for overlaps

nyalldawson commented on tests/src/python/test_qgsrange.py:
```diff
@@ -212,14 +214,16 @@ def testOverlaps(self):
         self.assertTrue(range.overlaps(QgsIntRange(1, 11)))
         self.assertTrue(range.overlaps(QgsIntRange(0, 9)))
-        self.assertTrue(range.overlaps(QgsIntRange(0, 10)))
+        self.assertFalse(range.overlaps(QgsIntRange(0, 10)))
```
This is wrong -- these two ranges DO overlap

Djedouas replied:
That's right, I have to fix this

Djedouas commented:
Changed the logic of the `overlaps()` method to simplify its code.

Changed the test, to test every combination of range positions and include/exclude bounds.

Djedouas committed:
fix(qgsrange): cppcheck wants a return path

nyalldawson commented on tests/src/python/test_qgsrange.py:
```diff
@@ -165,69 +165,114 @@ def testContainsElement(self):
+        # reference range
+        refLower, refUpper = 0, 10
+
+        # other range is compeltely before or after reference range
```
```suggestion
        # other range is completely before or after reference range
```

nyalldawson commented on tests/src/python/test_qgsrange.py:
```diff
@@ -165,69 +165,114 @@ def testContainsElement(self):
+            refIncUpper,
+            expected,
+        ) in [
+            # referance range and other range are contiguous
```
```suggestion
            # reference range and other range are contiguous
```

nyalldawson commented on tests/src/python/test_qgsrange.py:
```diff
@@ -165,69 +165,114 @@ def testContainsElement(self):
+            [3, 10],
+            # same bounds
+            [0, 10],
+            # reference range is compeltely inside other range
```
```suggestion
            # reference range is completely inside other range
```

nyalldawson commented on tests/src/python/test_qgsrange.py:
```diff
@@ -165,69 +165,114 @@ def testContainsElement(self):
+            [0, 10],
+            # reference range is compeltely inside other range
+            [-2, 14],
+            # other range is compeltely inside reference range
```
```suggestion
            # other range is completely inside reference range
```

nyalldawson commented on src/core/qgsrange.h:
```diff
@@ -574,26 +570,26 @@ class QgsTemporalRange
     bool overlaps( const QgsTemporalRange<T> &other ) const
     {
-      if ( !mUpper.isValid() && ( ( mIncludeLower && mLower <= other.mUpper ) || ( !mIncludeLower && mLower < other.mUpper ) ) )
+      if ( !mUpper.isValid() && ( ( mIncludeLower && other.mIncludeUpper && mLower <= other.mUpper ) || ( ( !mIncludeLower || !other.mIncludeUpper ) && mLower < other.mUpper ) ) )
```
Shouldn't this method be synchronized with QgsRange::overlaps logic?

Djedouas replied:
In fact, going by QgsRange::overlaps new logic, I found that QgsTemporalRange::overlaps was not considering infinite bounds (invalid QDate()) so we had

```python
r1 = QgsDateRange(QDate(), QDate(2000, 1, 1))
r2 = QgsDateRange(QDate(3000, 1, 1), QDate())
r1.overlaps(r2)
>> True
```

Now with the new logic we are good (hopefully) with all the combinations I could add to the tests.

Thanks for the review!

Djedouas committed:
fix(qgsdaterange): consider infinite bounds in overlaps method

Infinite left or right date bound were not considered when computing
overlaps, leading to ranges being considered overlapping where they
don't.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

nyalldawson commented on src/core/qgsrange.h:
```diff
@@ -571,31 +567,47 @@ class QgsTemporalRange
     bool overlaps( const QgsTemporalRange<T> &other ) const
     {
-      if ( !mUpper.isValid() && ( ( mIncludeLower && mLower <= other.mUpper ) || ( !mIncludeLower && mLower < other.mUpper ) ) )
+      // one of both range is infinite
```
```suggestion
      // one or both range is infinite
```

nyalldawson committed:
Update src/core/qgsrange.h
